### DISCUSSION TITLE
Compiling assembly files with different arguments

### DISCUSSION THREAD

**Discussion by gitkarokan** [mesonbuild/meson#10497] [Q&A] [answered]
Hello,

I'm setting up a new build system that uses GCC as compiler, assembler, and linker.

I'm using a cross_file.build to set up the general compilation flags for C files ( c_args) and linker (c_link_args).
The options for assembly files are different from C files which is why I was trying to set it up in the cross-compilation file, however, I cannot find the keyword for assembly flags (I tried with a_flags, s_flags, etc... ) but I get the ERROR: Unknown options:

Do you know what should be the keyword for assembly flags?

I'm also seeing that the command lines have extra flags from the one I defined, how can I force to only use the one defined in the cross-compilation file?

Thanks in advance!

**Comment by eli-schwartz** [comment]
asm `*.s` files are currently handled by marking the C compiler as supporting it as a possible file extension, and relying on GCC detecting that it is asm and dispatching to the assembler. So the correct option is c_args. You can pass asm-specific compiler options with `-Wa,option,option,option`. This should be ignored for `*.c` sources.

**Comment by gitkarokan** [comment]
Thanks a lot @eli-schwartz for your quick feedback!

Please see below the command line to compile an assembly file with the build chain from ST - SPC5Studio 

`ppc-freevle-eabi-gcc -x assembler-with-cpp -c -mcpu=e200z4 -mvle -meabi -msdata=none -mregnames -Wa,-amhls=build/lst/boot.lst -DPPC_USE_VLE=1  -MD -MP -MF .dep/boot.o.d -I. -I./components/spc58xgxx_platform_component_rla/lib/include -I./components/spc58xgxx_platform_component_rla/cfg -I./components ./components/spc58xgxx_platform_component_rla/lib/src/boot.s -o build/obj/boot.o
`
I removed the -c option as you suggested earlier so that meson can detect the compiler, however there are few challenges with the command line I hope you can help me.

1) if I add `-x assembler-with-cpp`  to my cross-compilation file, I'm able to run the meson configuration, however, this will be included in all the C files compilation and the compiler does not know how to interpret it.

I tried adding the `-Wa,-x  assembler-with-cpp` and `-Wa,-x ,assembler-with-cpp`  however in this case meson fails to verify the compiler, do you have any suggestions?

2) the compiler is invoked with source file name dependency, for example -Wa,-amhls=build/lst/`boot.lst` which means the flag should change with each source file for example for ctr0.s file the command should be something like -Wa,-amhls=build/lst/`ctr0.lst` , in scons I was able to modify the flag based on the source to be compiled, do we have similar tools in meson?

Thanks a lot!

----
Cross Compile file

```
[constants]
compiler_path = '<global_path_to_compiler>'
arch = 'ppc-freevle-eabi-' 

[binaries]
c       = compiler_path + arch + 'gcc'
cpp     = compiler_path + arch + 'cpp'
ld      = compiler_path + arch + 'gcc'
ar      = compiler_path + arch + 'ar'
as      = compiler_path + arch + 'as'
size    = compiler_path + arch + 'size'
objdump = compiler_path + arch + 'objdump'
objcopy = compiler_path + arch + 'objcopy'
strip   = compiler_path + arch + 'strip'
gdb     = compiler_path + arch + 'gdb'

[built-in options]
c_args      = [     
                #'-Wa,-x,assembler-with-cpp',
                '-mcpu=e200z4',
                '-mvle',
                '-meabi',
                '-msdata=none',
                '-mregnames',
                '-O2',
                '-gdwarf-2',
                '-fomit-frame-pointer',
                '-falign-functions=16',
                '-fno-gcse',
                '-std=gnu99',
                '-DPPC_USE_VLE=1',
                '-MD',
                '-MP'
               ]

c_link_args = ['-mcpu=e200z4',
               '-mvle',
               '-meabi',
               '-msdata=none',
               '-mregnames',
               '-O2',
               '-gdwarf-2',
               '-fomit-frame-pointer',
               '-falign-functions=16',
               '-fno-gcse',
               '-std=gnu99',
               '-ffunction-sections',
               '-fdata-sections',
               '-fno-common',
               '-nostartfiles',
               '-Wl,-Map=out.map,--cref,--no-warn-mismatch,--library-path=,--gc-sections,-lm,-lc,-z,max-page-size=16384,--script=../application.ld',
               '-o',
               'out.elf'
              ]

[properties]
# Directory that contains 'bin', 'lib', etc
root = '.\tools\gcc_compiler'
# Directory that contains 'bin', 'lib', etc for the toolchain and system libraries
sys_root = compiler_path

[host_machine]
system = 'bare metal'
cpu_family = 'ppc'
cpu = 'e200z4'
endian = 'little'
```

**Comment by eli-schwartz** [comment, marked as answer]
> I tried adding the -Wa,-x assembler-with-cpp and -Wa,-x ,assembler-with-cpp however in this case meson fails to verify the compiler, do you have any suggestions?

I think I misunderstood what you want... `-Wa` says "pass this option to the assembler". `-x` says "in the frontend, interpret this file as $language, in this case assembly, and invoke the assembler".

So you cannot pass -x via -Wa, because the point of -x is to decide whether to use -Wa.

The `-x` option allows you to override the compiler's automatic detection of the language based on the file extension. It's utterly inappropriate to include in a cross file, because if it is correct to pass that option, it must be defined in meson.build as part of the project, not as part of an optional environment.

More generally... why do you want to pass -x at all? Can't you just name your source files `*.S`, which will be autodetected as assembler-with-cpp?

**Reply by eli-schwartz** [reply]
Alright, it seems this is basically just #10431

As mentioned there, asm support should likely be factored out into its own language, e.g. #9379

Also mentioned there:
> According to the GNU documentation, all modern GCC do it (which is nice, since `-x assembler-with-cpp` is no longer needed for `.S`),

So maybe just remove that? :)

**Reply by gitkarokan** [reply]
Thank you @eli-schwartz for the detailed explanation, you are correct I was mixing -x and -Wa and sorry for missing #10431 

I rename the file to be *.S that works, Thanks!

I would prefer to have a separate flag for the assembler so in that way I would have more control

Again, thank you!
